lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/driver/AzureBatchTaskStatusAlarmHandler.java

```diff
@@ -0,0 +1,150 @@
+  private final Clock clock;
+
+  private static final Logger LOG = Logger.getLogger(AzureBatchTaskStatusAlarmHandler.class.getName());
+
+  @Inject
+  private AzureBatchTaskStatusAlarmHandler(
+      final InjectionFuture<AzureBatchEvaluatorShimManager> evaluatorShimManager,
+      final AzureBatchHelper azureBatchHelper,
+      final Evaluators evaluators,
+      final Clock clock,
+      @Parameter(AzureBatchTaskStatusCheckPeriod.class) final int taskStatusCheckPeriod) {
+    this.evaluatorShimManager = evaluatorShimManager;
+    this.azureBatchHelper = azureBatchHelper;
+    this.evaluators = evaluators;
+    this.clock = clock;
+    this.taskStatusCheckPeriod = taskStatusCheckPeriod;
+  }
+
+  /**
+   * This method is periodically invoked by the Runtime Clock. It will call Azure Batch APIs to determine
+   * the status of tasks running inside the job and notify REEF of tasks statuses that correspond to running
+   * evaluators.
+   *
+   * @param alarm alarm object.
+   */
+  @Override
+  public void onNext(final Alarm alarm) {
+    String jobId = this.azureBatchHelper.getAzureBatchJobId();
+    List<CloudTask> allTasks = this.azureBatchHelper.getTaskStatusForJob(jobId);
+
+    // Report status if the task has an associated active container.
+    LOG.log(Level.FINER, "Found {0} tasks from job id {1}", new Object[]{allTasks.size(), jobId});
+    for (CloudTask task : allTasks) {
+      Optional<EvaluatorManager> optionalEvaluatorManager = this.evaluators.get(task.id());
+
+      if (!optionalEvaluatorManager.isPresent() && !TaskState.COMPLETED.equals(task.state())) {
+        // This usually means that the evaluator shim has started, but hasn't sent the status message
+        // back to the driver yet.
+        LOG.log(Level.FINE, "No Evaluator found for Azure Batch task id = {0}. Ignoring.", task.id());
+      } else if (!optionalEvaluatorManager.isPresent() && TaskState.COMPLETED.equals(task.state())) {
+        // This indicates that the evaluator shim exited prematurely. We inform REEF of resource allocation
+        // so it's possible to trigger an event signaling resource failure later.
+        LOG.log(Level.INFO, "Azure Batch task id = {0} is in 'COMPLETED' state, but it does not have " +
+            "an Evaluator associated with it. This indicates that the evaluator shim has failed before " +
+            "it could send a callback to the driver.", task.id());
+        this.evaluatorShimManager.get().onResourceAllocated(task.id(), PLACEHOLDER_REMOTE_ID, Optional.of(task));
+      } else if (optionalEvaluatorManager.get().isClosedOrClosing()) {
+        LOG.log(Level.FINE, "Evaluator id = {0} is closed. Ignoring.", task.id());
+      } else {
+        LOG.log(Level.FINE, "Reporting status for Task Id: {0} is [Azure Batch Status]:{1} ",
+            new Object[]{task.id(), task.state().toString()});
+        this.evaluatorShimManager.get().onAzureBatchTaskStatus(task);
+      }
+    }
+
+    if (this.isAlarmEnabled()) {
```

Comment:
Is there a possible race condition here? Could it be the case that this.isAlarmEnabled() returns true, then before this.scheduleAlarm() is called, another thread calls disableAlarm()?

Reply:
answer for line 143 together: we addressed the race condition problem by having the "synchronize" keyword in enableAlarm(), diusableAlarm(), isAlarmEnabled(), as "synchronize" locked on the class instance.

Reply:
I wonder if it is possible for a race condition to occur between the call to this.isAlarmEnabled() and this.scheduleAlarm()? Between these two calls in the onNext method, the synchronize lock on the class instance is not held, thereby allowing the possibility of scheduling an alarm when this.isAlarmEnabled() == false due to some other thread calling disableAlarm() in the interim.

Reply:
Good point. I think that is possible. What do you think changing line 118 to:

```
    synchronized (this) {
      if (this.isAlarmEnabled()) {
        this.scheduleAlarm();
      }
    }
```

Reply:
Yes, that is what I would do, and no need for synchronize in private methods e.g., isAlarmEnabled()

Reply:
Thanks. Uploaded a fix as agreed.